pkg/cloudprovider/provider/vsphere/helper.go

```diff
@@ -106,21 +106,48 @@ func CreateLinkClonedVm(vmName, vmImage, datacenter, clusterName, folder string,
 	if containerLinuxUserdata != "" {
 		userdataBase64 := base64.StdEncoding.EncodeToString([]byte(containerLinuxUserdata))
 
-		//Found out via `govc  vm.info -json containerlinux-template`
-		//Both `Key` and `Id` must be correct in order for this to work
-		//You will however not get an error from the API if its wrong
-		//Only the `Id` is part of the source OVA image, the `Key` gets generated by Vsphere
-		//and I have no idea if that generation is deterministic
-		vAppAconfig = &types.VmConfigSpec{Property: []types.VAppPropertySpec{
-			{
-				Info:            &types.VAppPropertyInfo{Key: 7, Id: "guestinfo.coreos.config.data.encoding", Value: "base64"},
-				ArrayUpdateSpec: types.ArrayUpdateSpec{Operation: types.ArrayUpdateOperationEdit},
-			},
-			{
-				Info:            &types.VAppPropertyInfo{Key: 4, Id: "guestinfo.coreos.config.data", Value: userdataBase64},
-				ArrayUpdateSpec: types.ArrayUpdateSpec{Operation: types.ArrayUpdateOperationEdit},
-			},
-		}}
+		// The properties describing userdata will already exist in the CoreOS VM template.
+		// In order to overwrite them, we need to specify their numeric Key values,
+		// which we'll extract from that template.
+		var mvm mo.VirtualMachine
+		err = templateVm.Properties(ctx, templateVm.Reference(), []string{"config", "config.vAppConfig", "config.vAppConfig.property"}, &mvm)
+		if err != nil {
+			return err
+		}
+
+		var property []types.VAppPropertySpec
+		info := mvm.Config.VAppConfig.GetVmConfigInfo()
+		properties := info.Property
+		for _, item := range properties {
+			switch item.Id {
+			case "guestinfo.coreos.config.data":
+				property = append(property, types.VAppPropertySpec{
+					ArrayUpdateSpec: types.ArrayUpdateSpec{
```

Comment:
just wanted to see if this doesn't have to be a pointer. the source code looks really ugly.

```
type ArrayUpdateSpec struct {

   DynamicData
    
   Operation ArrayUpdateOperation `xml:"operation"`
   RemoveKey AnyType              `xml:"removeKey,omitempty,typeattr"`
   }
    
   func init() {
   t["ArrayUpdateSpec"] = reflect.TypeOf((*ArrayUpdateSpec)(nil)).Elem()
   }
```

Reply:
Yeah, the vmware sdk is quite special...